ecl/hql/hqlopt.cpp

```diff
@@ -4269,9 +4269,36 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
 
                         if (ok)
                         {
-                            //If expanding the project removed all references to left (very silly join....) make it an all join
-                            if (transformed->hasAttribute(lookupAtom) && !exprReferencesDataset(&args.item(2), newLeft))
+                            //If expanding the project removed all references to left (very silly join....)
+                            //then it can be converted to a an all join against a filtered right dataset
+                            IHqlExpression * transformedCond = &args.item(2);
+                            if (transformed->hasAttribute(lookupAtom) && !exprReferencesDataset(transformedCond, newLeft))
+                            {
+                                IHqlExpression * rightDs = &args.item(1);
+                                OwnedHqlExpr right = createSelector(no_right, rightDs, transformedSeq);
+                                if (exprReferencesDataset(transformedCond, right))
+                                {
+                                    OwnedHqlExpr newCond = replaceSelector(transformedCond, right, rightDs);
+                                    OwnedHqlExpr filteredRightDs = createDataset(no_filter, { LINK(rightDs), LINK(newCond) });
+                                    //Replace the right dataset with the filtered dataset, and the join condition is now unconditional
+                                    args.replace(*filteredRightDs.getClear(), 1);
+                                    args.replace(*createConstant(true), 2);
+                                }
+
+                                //This test currently duplicates the test in the enclosing if.  This is in preparation for applying the logic
+                                //to other joins later by removing the outer test.
+                                if (transformed->hasAttribute(lookupAtom))
+                                {
+                                    //LOOKUP join implicitly dedups the right hand side by the hard-match-condition (unless MANY is specified).
+                                    //Technically this is not equivalent to KEEP(1) - since that is only applied after any soft-match condition
+                                    //However in this case the match condition is true (or row-invarient in very obscure cases),
+                                    //so they are identical and this transformation is valid.
+                                    if (!transformed->hasAttribute(manyAtom))
+                                        args.append(*createAttribute(keepAtom, createConstant(1)));
```

Comment:
Ithink this is correct, but hesitated. JOIN,LOOKUP only supports support KEEP if MANY, so there cannot be an pre-existing KEEP if !many.